datafusion/expr/src/udf.rs

```diff
@@ -420,6 +425,11 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     fn short_circuits(&self) -> bool {
         false
     }
+
+    /// Coerce the types of the arguments to the types that the function expects
```

Comment:
Nice -- I think it would be nice to expand this documentation to help people understand the API some more

Perhaps something like 

```suggestion
    /// Coerce arguments of a function call to types that the function can evaluate.
    ///
    /// This function is only called if [`Self::signature`] returns [`TypeSignature::UserDefined`]. Most
    /// UDFs should return one of the other variants of `TypeSignature` which handle common
    /// cases
    ///
    /// See the [type coercion module](datafusion_expr::type_coercion)
    /// documentation for more details on type coercion
    ///
    /// For example, if your function requires a floating point arguments, but the user calls
    /// it like `my_func(1::int)` (aka with `1` as an integer), coerce_types could return `[DataType::Float64]`
    /// to ensure the argument was cast to `1::double`
    ///
    /// # Parameters
    /// * `arg_types`: The argument types of the arguments  this function with
    ///
    /// # Return value
    /// A Vec the same length as `arg_types`. DataFusion will `CAST` the function call
    /// arguments to these specific types. 
```